src/fpm/downloader.f90

```diff
@@ -0,0 +1,94 @@
+module fpm_downloader
+  use fpm_error, only: error_t, fatal_error
+  use fpm_filesystem, only: which
+  use fpm_versioning, only: version_t
+  use jonquil, only: json_object, json_value, json_error, json_load, cast_to_object
+
+  implicit none
+  private
+
+  public :: downloader_t
+
+  !> This type could be entirely avoided but it is quite practical because it can be mocked for testing.
+  type downloader_t
+  contains
+    procedure, nopass :: get_pkg_data, get_file, unpack
+  end type
+
+contains
+
+  !> Perform an http get request and save output to file.
+  subroutine get_pkg_data(url, version, tmp_file, json, error)
+    character(*), intent(in) :: url
+    type(version_t), allocatable, intent(in) :: version
+    character(*), intent(in) :: tmp_file
+    type(json_object), intent(out) :: json
+    type(error_t), allocatable, intent(out) :: error
+
+    class(json_value), allocatable :: j_value
+    type(json_object), pointer :: ptr
+    type(json_error), allocatable :: j_error
+
+    if (allocated(version)) then
+      ! Request specific version.
+      call get_file(url//'/'//version%s(), tmp_file, error)
+    else
+      ! Request latest version.
+      call get_file(url, tmp_file, error)
+    end if
+    if (allocated(error)) return
+
+    call json_load(j_value, tmp_file, error=j_error)
+    if (allocated(j_error)) then
+      allocate (error); call move_alloc(j_error%message, error%message); call json%destroy(); return
+    end if
+
+    ptr => cast_to_object(j_value)
+    if (.not. associated(ptr)) then
+      call fatal_error(error, "Error parsing JSON from '"//url//"'."); return
+    end if
+
+    json = ptr
+  end
+
+  subroutine get_file(url, tmp_file, error)
```

Comment:
If you only look at the procedure name, then yes, you are right. But this is a member procedure, meaning that it will, for example, be used as `downloader%get_file()`. Therefore I think it will always be clear that this procedure downloads a file from the web.

Reply:
It's been changed here: https://github.com/fortran-lang/fpm/pull/817/commits/0c590794729fc8a859c9dfca2a9f5a886897ec75